File: backend/app/schemas/lighting.py
"""Pydantic schemas for Lighting Calibration module."""
from __future__ import annotations
from typing import Optional
from pydantic import BaseModel


class BrightnessStats(BaseModel):
    mean: float
    median: float
    min_val: float
    max_val: float
    std: float
    pct_of_full_scale: float
    status: str  # pass / warning / fail


class HistogramResult(BaseModel):
    bins: list[float]        # 256 bin centres
    counts: list[float]      # normalised % per bin
    underexposed_pct: float
    overexposed_pct: float
    clipping_low: bool
    clipping_high: bool
    peak_bin: int


class UniformityResult(BaseModel):
    uniformity_min_max: float   # Min/Max × 100
    uniformity_cv: float        # (1 – Std/Mean) × 100
    status: str                 # excellent / good / acceptable / bad
    heatmap_data: list[list[float]]
    heatmap_rows: int
    heatmap_cols: int


class HotspotInfo(BaseModel):
    x: int
    y: int
    width: int
    height: int
    severity: float
    type: str   # bright / shadow / gradient


class HotspotResult(BaseModel):
    hotspots: list[HotspotInfo]
    severity_score: float
    heatmap_data: list[list[float]]
    heatmap_rows: int
    heatmap_cols: int


class NoiseResult(BaseModel):
    spatial_mean: float
    spatial_std: float
    temporal_mean: float
    temporal_std: float
    noise_map: list[list[float]]
    noise_map_rows: int
    noise_map_cols: int


class SNRResult(BaseModel):
    snr_linear: float
    snr_db: float
    status: str  # excellent / good / poor


class FlickerResult(BaseModel):
    brightness_over_time: list[float]
    frame_indices: list[int]
    mean: float
    std: float
    has_flicker: bool
    flicker_pct: float
    frequency_estimate: Optional[float]
    fft_frequencies: list[float]
    fft_amplitudes: list[float]


class DynamicRangeResult(BaseModel):
    effective_dr_stops: float
    usable_dr_stops: float
    signal_mean: float
    noise_floor: float


class LightingAnalysisResult(BaseModel):
    num_frames: int
    image_shape: list[int]
    bit_depth: int
    brightness: BrightnessStats
    histogram: HistogramResult
    uniformity: UniformityResult
    hotspots: HotspotResult
    noise: NoiseResult
    snr: SNRResult
    flicker: Optional[FlickerResult]
    dynamic_range: Optional[DynamicRangeResult]
    recommendations: list[str]
    overall_status: str  # pass / warning / fail

File: frontend/src/types/lighting.ts
// TypeScript types mirroring backend Lighting schemas

export interface BrightnessStats {
  mean: number
  median: number
  min_val: number
  max_val: number
  std: number
  pct_of_full_scale: number
  status: 'pass' | 'warning' | 'fail'
}

export interface HistogramResult {
  bins: number[]
  counts: number[]
  underexposed_pct: number
  overexposed_pct: number
  clipping_low: boolean
  clipping_high: boolean
  peak_bin: number
}

export interface UniformityResult {
  uniformity_min_max: number
  uniformity_cv: number
  status: 'excellent' | 'good' | 'acceptable' | 'bad'
  heatmap_data: number[][]
  heatmap_rows: number
  heatmap_cols: number
}

export interface HotspotInfo {
  x: number
  y: number
  width: number
  height: number
  severity: number
  type: 'bright' | 'shadow' | 'gradient'
}

export interface HotspotResult {
  hotspots: HotspotInfo[]
  severity_score: number
  heatmap_data: number[][]
  heatmap_rows: number
  heatmap_cols: number
}

export interface NoiseResult {
  spatial_mean: number
  spatial_std: number
  temporal_mean: number
  temporal_std: number
  noise_map: number[][]
  noise_map_rows: number
  noise_map_cols: number
}

export interface SNRResult {
  snr_linear: number
  snr_db: number
  status: 'excellent' | 'good' | 'poor'
}

export interface FlickerResult {
  brightness_over_time: number[]
  frame_indices: number[]
  mean: number
  std: number
  has_flicker: boolean
  flicker_pct: number
  frequency_estimate: number | null
  fft_frequencies: number[]
  fft_amplitudes: number[]
}

export interface DynamicRangeResult {
  effective_dr_stops: number
  usable_dr_stops: number
  signal_mean: number
  noise_floor: number
}

export interface LightingAnalysisResult {
  num_frames: number
  image_shape: number[]
  bit_depth: number
  brightness: BrightnessStats
  histogram: HistogramResult
  uniformity: UniformityResult
  hotspots: HotspotResult
  noise: NoiseResult
  snr: SNRResult
  flicker: FlickerResult | null
  dynamic_range: DynamicRangeResult | null
  recommendations: string[]
  overall_status: 'pass' | 'warning' | 'fail'
}
